### PR TITLE
2447 fix pgvector tests and notebook

### LINKED ISSUE #2447: Fix pgvector tests and notebook
@Knucklessg1 , could you please fix the pgvector's notebook and tests? Thanks.

- The notebook has not been ran. I'd suggest taking the first two examples from `RetrieveChat` notebook instead of using all of them, and actually run the notebook.
- As I tested locally, the current `RetrieveChatTest` would skip tests for pgvector as `from autogen.agentchat.contrib.vectordb.pgvector import PGVector` should be `from autogen.agentchat.contrib.vectordb.pgvectordb import PGVectorDB`, other contents should be corrected as well. And I see `AttributeError: 'NoneType' object has no attribute 'cursor'` after I corrected the import and naming.
- The `Install and Start PostgreSQL` in CI doesn't work for me either.

_Originally posted by @thinkall in https://github.com/microsoft/autogen/issues/2439#issuecomment-2066277220_

### PR DESCRIPTION
**Pull request by Knucklessg1**
## Why are these changes needed?

- Notebook was updated with missing details.
- Corrected pytests
- Added client error handling
- Fixed missing username and password parameters
- Fixed URL encoding of connection string to support symbols in password and username
- Fixed adding documents logic
- Added the ability to choose the sentence transformer model.
- Fixed pre-commits - Bumped openai minimum version to 1.23.3 to support OpenAI vector_store

## Related issue number

Closes #2447 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by thinkall** [comment]
I pushed some fixes:
- run notebook
- improve db.get_collection
- fix db.delete_collection

But in the `Collection` class,  more fixes are needed.  The tests can't pass.

**Comment by thinkall** [comment]
Why none of `ContribTests` is triggered? @ekzhu , @sonichi ,  @jackgerrits any thought here?

**Review by thinkall — CHANGES_REQUESTED** [review]
Tests need to be fixed.

File: setup.py
Comment (Knucklessg1) [inline setup.py:54]: @thinkall  this change was made because there is a hard dependency on chromadb through retrievechat_user_proxy and retrieve-utils. Until that dependency is removed, we also need to install chromadb with qdrant and pgvector.

File: notebook/agentchat_function_call_code_writing.ipynb
Comment (Knucklessg1) [reply notebook/agentchat_function_call_code_writing.ipynb]: Removed this. Not sure how that happened.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (thinkall) [inline autogen/agentchat/contrib/vectordb/pgvectordb.py:847]: According to the protocol in base, if ids is None, all docs should be returned.

File: setup.py
Comment (thinkall) [reply setup.py:54]: OK. I'd prefer remove chromadb from the `retrieve_chat` list and add it to each of the extras, as ideally, a db should not depend on another db.

File: test/agentchat/contrib/vectordb/test_pgvectordb.py
Comment (thinkall) [inline test/agentchat/contrib/vectordb/test_pgvectordb.py:86]: The format of `res` doesn't meet the protocol.

File: test/agentchat/contrib/vectordb/test_pgvectordb.py
Comment (thinkall) [inline test/agentchat/contrib/vectordb/test_pgvectordb.py:96]: if no ids is given, return all docs.

**Review by thinkall — CHANGES_REQUESTED** [review]

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (thinkall) [inline autogen/agentchat/contrib/vectordb/pgvectordb.py:848]: Notebook also fails.

**Review by thinkall — CHANGES_REQUESTED** [review]

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py:848]: What is the expected behavior when IDs list is None or the list length is 0?

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py:847]: Understood, let me correct

Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py]: This was corrected. All Docs will be returned. Pytest test case also added to validate all docs are returned.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py]: Should be fixed now.

File: test/agentchat/contrib/vectordb/test_pgvectordb.py
Comment (Knucklessg1) [reply test/agentchat/contrib/vectordb/test_pgvectordb.py:86]: What is the expectation? Are you looking for an exact data format? 

assert final_results == ["3", "2"]?

File: test/agentchat/contrib/vectordb/test_pgvectordb.py
Comment (Knucklessg1) [reply test/agentchat/contrib/vectordb/test_pgvectordb.py]: Added this test case.

File: test/agentchat/contrib/vectordb/test_pgvectordb.py
Comment (Knucklessg1) [reply test/agentchat/contrib/vectordb/test_pgvectordb.py]: Resolved

**Review by thinkall — APPROVED** [review]
Thank you @Knucklessg1 !

**Comment by Knucklessg1** [comment]
@sonichi these changes have been validated. Can you assist with final review?

**Review by Hk669 — APPROVED** [review]
lgtm.

File: .github/workflows/contrib-openai.yml
Comment (Hk669) [inline .github/workflows/contrib-openai.yml:33]: should be added to the secrets.

File: .github/workflows/contrib-tests.yml
Comment (Hk669) [inline .github/workflows/contrib-tests.yml:78]: here too.

File: .github/workflows/contrib-openai.yml
Comment (Knucklessg1) [reply .github/workflows/contrib-openai.yml:33]: Who can help add these to secrets?

Comment (sonichi) [reply .github/workflows/contrib-openai.yml:33]: I've added POSTGRES_PASSWORD to the secrets. You can use `${{ secrets.POSTGRES_PASSWORD }}`

File: setup.py
Comment (sonichi) [inline setup.py:17]: Is older openai version unsupported?

File: .github/workflows/contrib-tests.yml
Comment (sonichi) [reply .github/workflows/contrib-tests.yml:78]: If this is for local test only, maybe it's OK. Open to other opinions.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (giorgossideris) [inline autogen/agentchat/contrib/vectordb/pgvectordb.py:627]: Again here, it is possible to use a different model.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (giorgossideris) [inline autogen/agentchat/contrib/vectordb/pgvectordb.py:588]: Same as above here.

File: setup.py
Comment (Knucklessg1) [reply setup.py:17]: Yes, older openai version is now unsupported as a result of this merge: https://github.com/microsoft/autogen/pull/2466

This was caught from the pre-commit checks.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py:37]: I wasn't aware of this, let me update the link. Thank you.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py:123]: Maybe I am misunderstanding, but the if the user decides to supply the embedding_function such as: SentenceTransformer("MODEL_NAME"). Then it should not result in an error. 

They could also be able to supply the ChromaDB embedding_function as well and it should work. It's a drop-in replacement.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py]: Resolved with updated link.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py]: Resolved with updated link.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Knucklessg1) [reply autogen/agentchat/contrib/vectordb/pgvectordb.py:630]: Resolved by updating the link.

File: autogen/agentchat/contrib/vectordb/pgvectordb.py
Comment (Hk669) [inline autogen/agentchat/contrib/vectordb/pgvectordb.py]: please update the link. Thanks
```suggestion
                f"from https://huggingface.co/models?library=sentence-transformers\nError: {e}"
            )
```

**Review by Hk669 — APPROVED** [review]

**Review by ekzhu — APPROVED** [review]

File: notebook/agentchat_pgvector_RetrieveChat.ipynb
Comment (ekzhu) [inline notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: 1. Remove 

```python
config_list = autogen.config_list_from_json(env_or_file="OAI_CONFIG_LIST")
```
And replace it with a raw Python list of models. We are moving away from using `OAI_CONFIG_LIST` in docs. 

2. Use 

```python
chat_result = user_proxy.initiate_chat(...)
```

To avoid printing a large blob of chat_result in the cell output.

Comment (Knucklessg1) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: Manually created the config_list and chat_result captured for all initiat_chats. 

I am unable to re-run the notebooks however. Would someone please re-run them to reset the output?

Comment (thinkall) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: I've reran the notebook.

Comment (thinkall) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: @ekzhu , `And replace it with a raw Python list of models.` the raw list increased the risk of API key leaking and the effort to clean the notebook before push and added extra steps before it can be ran.

Comment (ekzhu) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: We can use `os.getenv("OPENAI_API_KEY")`

Comment (thinkall) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: > We can use `os.getenv("OPENAI_API_KEY")`
> 
> 

But it doesn't work for Azure OpenAI, nor combination of different configs. What's the reason for removing config list file?

**Comment by ekzhu** [comment]
@sonichi check added dependencies in setup.py for tests.

File: setup.py
Comment (sonichi) [reply setup.py:17]: This is an issue. I think #2466 was meant to support both old and new openai versions. Is it possible to support both? I don't think every user is ready to migrate to the new openai version.
cc @IANTHEREAL

File: setup.py
Comment (sonichi) [inline setup.py:38]: "chromadb" is not supposed to be in the list. Maybe there's a reason and I missed it.

Comment (Knucklessg1) [reply setup.py:38]: chromadb is a hard dependency for all RAG agent types because Chroma is coupled tightly with RAG agent type. Once we remove chromaDB as a hard dependency, we can remove it from base retrieve chat requirements. Until then, when qdrant or pgvector is installed, it also needs chromadb

Comment (Knucklessg1) [reply setup.py:38]: As for the open ai version, when on the older version, mypy throws an error that it is unable to find openAI vector_stores which is only available in latest (1.23.3)+

Comment (sonichi) [reply setup.py:38]: Thanks. We can keep the chromadb. For OpenAI version, we need to figure out how to support older versions.

Comment (Knucklessg1) [reply setup.py:38]: Should that be scoped for this PR? Or does that need to be addressed as a separate issue and another PR? I'm not sure how to approach the backwards compatibility requirement.

Comment (sonichi) [reply setup.py:38]: @IANTHEREAL is it true that openai<1.23.3 support is broken by #2466 ? Can we fix it easily? I agree that the fix is out of the scope of this PR.

Comment (Hk669) [reply setup.py:38]: > @IANTHEREAL is it true that openai<1.23.3 support is broken by #2466 ? Can we fix it easily? I agree that the fix is out of the scope of this PR. Is it OK to merge this PR and then fix the problem? Or is it better to fix that problem first?

@sonichi in the PR #2466 , the previous versions <1.21 has been removed. maybe adding them back would fix the problem. i would prefer, fix that problem and then merge this PR.

Comment (sonichi) [reply setup.py:38]: Thanks @Hk669 . @Knucklessg1 For a good user experience of backward compatibility, can we find a different solution to the mypy check error instead of increasing the required openai version? Could you share which lines fail the mypy check so that we can take a look how to resolve it?

Comment (thinkall) [reply setup.py:38]: maybe we can remove openai utils from the mypy check list for now and add it back when the compatibility issue fixed. @sonichi

Comment (sonichi) [reply setup.py:38]: That works.

Comment (IANTHEREAL) [reply setup.py:38]: I must have forgotten to add #type: ignore; that should work.


image

Comment (sonichi) [reply setup.py:38]: @WaelKarkoub kindly offers addressing the remaining issues in #2490 
So merging the PR now.

**Comment by sonichi** [comment]
@LittleLittleCloud I saw that Dotnet Build is required before merging and blocks the PR. This PR shouldn't require that. I'm going to remove it from the required action for now. We can discuss later.

File: notebook/agentchat_pgvector_RetrieveChat.ipynb
Comment (sonichi) [reply notebook/agentchat_pgvector_RetrieveChat.ipynb:1]: @ekzhu @thinkall could you discuss this on discord? I think both of your opinions need to be considered on this matter. For example, a comment containing a link to using Azure OAI or combination of different configs can be included, or some other approach.
I'm resolving the conversation for now for merging this PR and unblock other PRs.

**Pull request merged**
